helix-term/src/commands/typed.rs

```diff
@@ -108,30 +114,115 @@ fn open(cx: &mut compositor::Context, args: &[Cow<str>], event: PromptEvent) ->
     }
 
     ensure!(!args.is_empty(), "wrong argument count");
+
+    fn open_file(cx: &mut compositor::Context, path: &Path, pos: Position) -> anyhow::Result<()> {
+        let _ = cx.editor.open(path, Action::Replace)?;
+        let (view, doc) = current!(cx.editor);
+        let pos = Selection::point(pos_at_coords(doc.text().slice(..), pos, true));
+        doc.set_selection(view.id, pos);
+        // does not affect opening a buffer without pos
+        align_view(doc, view, Align::Center);
+        Ok(())
+    }
+
+    fn glob(path: &Path) -> anyhow::Result<Glob> {
+        let path_str = path.to_str().context("invalid unicode")?;
+        GlobBuilder::new(path_str)
+            .literal_separator(true)
+            .empty_alternates(true)
+            .build()
+            .context("invalid glob")
+    }
+
     for arg in args {
         let (path, pos) = args::parse_file(arg);
-        let path = helix_stdx::path::expand_tilde(path);
-        // If the path is a directory, open a file picker on that directory and update the status
-        // message
-        if let Ok(true) = std::fs::canonicalize(&path).map(|p| p.is_dir()) {
-            let callback = async move {
-                let call: job::Callback = job::Callback::EditorCompositor(Box::new(
-                    move |editor: &mut Editor, compositor: &mut Compositor| {
-                        let picker = ui::file_picker(path.into_owned(), &editor.config());
-                        compositor.push(Box::new(overlaid(picker)));
-                    },
-                ));
-                Ok(call)
+        let path = helix_stdx::path::canonicalize(path);
+        if let Ok(metadata) = path.metadata() {
+            // Path exists
+            // Shortcut for opening a path without globbing
+            if metadata.is_dir() {
+                // If the path is a directory, open a file picker on that directory
+                let callback = async move {
+                    let call: job::Callback = job::Callback::EditorCompositor(Box::new(
+                        move |editor: &mut Editor, compositor: &mut Compositor| {
+                            let picker = ui::file_picker(path, &editor.config());
+                            compositor.push(Box::new(overlaid(picker)));
+                        },
+                    ));
+                    Ok(call)
+                };
+                cx.jobs.callback(callback);
+            } else {
+                open_file(cx, &path, pos)?;
+            }
+            continue;
+        }
+
+        let mut glob_set_builder = GlobSet::builder();
```

Comment:
This entire logic seems both confusing/buggy and backwards to me. I would think you should just:

* check if the oath with the given name exists (open it if it does)
* use the override matcher of ignore to ignore everything (just an ignore for `*`) and then whiteliste the specified glob pattern

Reply:
I don't see how I could use `ignore` for this. Maybe I am missing something though.

I tried something like the following code, but it goes through the whole file system and takes 15 seconds on my machine:

```rust
fn main() {
    let walker = ignore::WalkBuilder::new("/")
        .overrides(
            ignore::overrides::OverrideBuilder::new("")
                .add("/bacon.toml")
                .unwrap()
                .build()
                .unwrap(),
        )
        .build();

    for entry in walker {
        let Ok(entry) = entry else {
            continue;
        };
        if entry.metadata().unwrap().is_file() {
            println!("{}", entry.path().display());
        }
    }
}
```

It should only go through the root directory and stop because the glob starts with `/`. I tried to add `!*` but then it doesn't match anything anymore (also not when `bacon.toml` exists).

Anyway, I don't think that we need `ignore`. I wrote an optimized walker without `ignore` and ask for a new review :)